[FEAT] Add Ability to Eat Food & Drink Smoothies in Bulk

elias7905 opened the pull request:
# **Description**
# Big Shoutout to @Polysemouse for helping with the gameService coding for this PR!

Added Eat 10 Button
![image](https://github.com/sunflower-land/sunflower-land/assets/101262042/dfe33d85-c973-4c86-89ed-01f52806f0a6)
Eat 10 Button disabled when less than 10
![image](https://github.com/sunflower-land/sunflower-land/assets/101262042/25e3a3bd-2ef9-4f91-9237-3ed5e0ea6ec0)
Mobile View
![image](https://github.com/sunflower-land/sunflower-land/assets/101262042/89a36a40-13ec-4c56-b087-6873a1bf12f3)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Ran Local Host

# Checklist:

- [x] Title of the PR is relevant and is prefixed with [FEAT], [CHORE] or [FIX]
- [x] Screenshot if it includes any UI changes
- [x] I have read the contributing guidelines and agree to the T&Cs
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

joaofernandez commented:
Niice! :))

Do you think it should be good to have some margin between the buttons? 

It helps to avoid miss click

![image](https://github.com/sunflower-land/sunflower-land/assets/5870502/d98b5fbc-471b-412f-b16b-4cb134021404)

joaofernandez commented on src/features/island/bumpkin/components/Feed.tsx:
```diff
@@ -86,12 +88,29 @@ export const Feed: React.FC<Props> = ({ food, onFeed }) => {
+              >
+                {isJuice(selected.name) ? "Drink 10" : "Eat 10"}
+              </Button>
+              <Button
+                disabled={!inventory[selected.name]?.gt(0)}
+                onClick={() => {
+                  const count = inventory[selected.name]?.toNumber() ?? 0;
+                  feed(selected, count);
+                }}
```
```suggestion
              <Button
                className="mt-1"
                disabled={!inventory[selected.name]?.gt(0)}
                onClick={() => {
                  const count = inventory[selected.name]?.toNumber() ?? 0;
                  feed(selected, count);
                }}
```

joaofernandez commented on src/features/island/bumpkin/components/Feed.tsx:
```diff
@@ -86,12 +88,29 @@ export const Feed: React.FC<Props> = ({ food, onFeed }) => {
+              >
+                {isJuice(selected.name) ? "Drink 1" : "Eat 1"}
+              </Button>
+              <Button
+                disabled={!inventory[selected.name]?.gte(10)}
+                onClick={() => feed(selected, 10)}
+              >
+                {isJuice(selected.name) ? "Drink 10" : "Eat 10"}
+              </Button>
```
```suggestion
              <Button
                className="mt-1"
                disabled={!inventory[selected.name]?.gte(10)}
                onClick={() => feed(selected, 10)}
              >
                {isJuice(selected.name) ? "Drink 10" : "Eat 10"}
              </Button>
```

joaofernandez commented:
Niice! :))

Do you think it should be good to have some margin between the buttons? 

It helps to avoid miss click

![image](https://github.com/sunflower-land/sunflower-land/assets/5870502/d98b5fbc-471b-412f-b16b-4cb134021404)

Polysemouse commented:
Instead of using loop and calling the feed function multiple times, new state mutation events should be created instead.

elias7905 commented:
> Instead of using loop and calling the feed function multiple times, new state mutation events should be created instead.

@Polysemouse Do you mind showing me how I can do that? I'm still quite new to coding in general

Polysemouse commented:
PR is ready to be reviewed

elias7905 commented:
> PR is ready to be reviewed

Thank you so much @Polysemouse!

elias7905 commented:
Added 1 more commit to Add a Feed All button

Polysemouse commented on src/features/island/bumpkin/components/Feed.tsx:
```diff
@@ -69,45 +65,110 @@ export const Feed: React.FC<Props> = ({ food, onFeed }) => {
+                    disabled={inventoryFoodCount.lessThan(1)}
+                    onClick={() => feed(1)}
+                  >
+                    {`${feedVerb} ${1}`}
```
You can just do 

${feedVerb} 1
Instead of 
${feedVerb} ${1}

elias7905 replied:
> You can just do
> 
> ${feedVerb} 1 Instead of ${feedVerb} ${1}

Wait wut lol ok let me correct that.

adamhannigan commented:
@elias7905 Thanks for the contribution 🤗 

We've had a similar PR raised before that didn't manage to enter the game.

The main concern was adding functionality that didn't have a significant impact on majority of gameplay. Although a small change, each addition does add some complexity + maintenance to our system.

We discovered that most players 99%+ do not hoard their food to be eaten later and those that hoarded did not face much friction when they needed to eat their way through it.

I'd be curious what use cases/player feedback influenced this PR. If there is a strong enough player case we can forge ahead, otherwise I'd prefer to omit this change for simplicity 🙏

elias7905 commented:
> those that hoarded did not face much friction when they needed to eat their way through it.

I've received quite a fair bit of feedback regarding this and those that hoarded usually hoard hundreds to thousands of food, and they find it tedious to click one by one to feed. I've been in a similar situation where I hoarded hundreds of pumpkin soup and Bumpkin Broth and man it was tedious to spam that eat 1 button

elias7905 commented:
> those that hoarded did not face much friction when they needed to eat their way through it.

@adamhannigan 
I've received quite a fair bit of feedback regarding this and those that hoarded usually hoard hundreds to thousands of food, and they find it tedious to click one by one to feed. I've been in a similar situation where I hoarded hundreds of pumpkin soup and Bumpkin Broth and man it was tedious to spam that eat 1 button

Polysemouse commented:
I agree with @elias7905.  Hoarding of food tends to be more common among players who progressed in the game.

I also agree that having the eat all button simplifies the code than just having the feed 10 button, as we are able to remove some unnecessary logic and the feed all button is present.

spencerdezartsmith commented:
> I agree with @elias7905. Hoarding of food tends to be more common among players who progressed in the game.
> 
> I also agree that having the eat all button simplifies the code than just having the feed 10 button, as we are able to remove some unnecessary logic and the feed all button is present.

Do you think hoarding food is a state that progressed players find themselves in now or it is a pattern that they will continue to engage in?

I want to try to keep this in perspective. Having 5-10 even 20 players ask for this in discord because they find themselves in this state is very different than it being a common strategy for 75%+ players. 

There is no real strategy behind hoarding food. It can't be sold etc so what is the reason? I think having these buttons implies there is some reason why you would hoard large amounts of food. I also think it's cluttered and it doesn't read well eg Drink 100 Bumpkin Smoothie. This is not grammatically correct so it looks very unpolished.

I would prefer to not implement this at this time.

Polysemouse commented:
@spencerdezartsmith There are several reasons where players find themselves hoarding food.  Some of the reasons could be 
- Players stokpiling food for delivery orders
- Players want to want for future boosts to food XP or get existing boost SFTs (eg. Golden Spatula) in the future before eating all their food
- Players want to focus on their daily farming/cooking activities and not care about eating food.  They would eat all the food item at once especially when it leads to a bumpkin level up

I believe this can bring substantial quality of life improvement to players.

>I also think it's cluttered and it doesn't read well eg Drink 100 Bumpkin Smoothie. This is not grammatically correct so it looks very unpolished.
I'd like to point out that this observation can also be applied to the Sell All crops interface.

spencerdezartsmith commented:
> Although I think the Eat 10 button standalone had more complexity compared to Eat 10 + Eat All)

Thanks for that. Ok, I can understand some of those reasons. Based on your above comment are you saying that the "Eat All" alone would be fine?

Polysemouse commented:
> > Although I think the Eat 10 button standalone had more complexity compared to Eat 10 + Eat All)
> 
> Thanks for that. Ok, I can understand some of those reasons. Based on your above comment are you saying that the "Eat All" alone would be fine?

Some players might want to leave a few food items to fulfill future daily deliveries, therefore I thikn the Eat 10 button is still needed.

Keeping the Eat 1 / 10 / All buttons will make it consistent with the other UI like the Sell Crops UI.

> Although I think the Eat 10 button standalone had more complexity compared to Eat 10 + Eat All

When we added the Eat All alongside with the Eat 10 button, we were able to remove some extra logic that we added in previous commits
https://github.com/sunflower-land/sunflower-land/pull/2973/commits/be1370e2ea2b934500c9594b38be871ceb034583

spencerdezartsmith commented:
> > > Although I think the Eat 10 button standalone had more complexity compared to Eat 10 + Eat All)
> > 
> > 
> > Thanks for that. Ok, I can understand some of those reasons. Based on your above comment are you saying that the "Eat All" alone would be fine?
> 
> Some players might want to leave a few food items to fulfill future daily deliveries, therefore I thikn the Eat 10 button is still needed.
> 
> Keeping the Eat 1 / 10 / All buttons will make it consistent with the other UI like the Sell Crops UI.
> 
> > Although I think the Eat 10 button standalone had more complexity compared to Eat 10 + Eat All
> 
> When we added the Eat All alongside with the Eat 10 button, we were able to remove some extra logic that we added in previous commits [be1370e](https://github.com/sunflower-land/sunflower-land/commit/be1370e2ea2b934500c9594b38be871ceb034583)

Do you think we could remove the name of the item when confirming the action and instead show the icon to move away from the bad grammar 😅 ? (For crops as well as food)

Polysemouse commented:
I think it is important to show the name of the item since player won't always know what the icon mean.  

An ideal solution is to have both the item name and icon in the UI.

It would be better to draft another PR for this (and some mockups) as this is kind of out of the scope of this PR.

elias7905 commented:
> Do you think we could remove the name of the item when confirming the action and instead show the icon to move away from the bad grammar 😅 ? (For crops as well as food)

I can try to fix the Grammar for the food and drinks but it might require additional steps for stuff that isn't just a simple 's' at the back

elias7905 commented:
> > > > Although I think the Eat 10 button standalone had more complexity compared to Eat 10 + Eat All)
> > > 
> > > 
> > > Thanks for that. Ok, I can understand some of those reasons. Based on your above comment are you saying that the "Eat All" alone would be fine?
> > 
> > 
> > Some players might want to leave a few food items to fulfill future daily deliveries, therefore I thikn the Eat 10 button is still needed.
> > Keeping the Eat 1 / 10 / All buttons will make it consistent with the other UI like the Sell Crops UI.
> > > Although I think the Eat 10 button standalone had more complexity compared to Eat 10 + Eat All
> > 
> > 
> > When we added the Eat All alongside with the Eat 10 button, we were able to remove some extra logic that we added in previous commits [be1370e](https://github.com/sunflower-land/sunflower-land/commit/be1370e2ea2b934500c9594b38be871ceb034583)
> 
> Do you think we could remove the name of the item when confirming the action and instead show the icon to move away from the bad grammar 😅 ? (For crops as well as food)

@spencerdezartsmith If it makes it less complex, I can push a commit that removes the Eat All button, so it's just Eat 1 and Eat 10, when less than 10 they can just eat 1 at a time

spencerdezartsmith commented:
Thanks @elias7905 I think we will leave this till we finish the upcoming season work. We can revisit then when we have more capacity to implement BE changes if we decide to go ahead with the change 🙏🏼

elias7905 commented:
> Thanks @elias7905 I think we will leave this till we finish the upcoming season work. We can revisit then when we have more capacity to implement BE changes if we decide to go ahead with the change 🙏🏼 
> 

@spencerdezartsmith No worries, I understand that the upcoming season and fishing updates priority

adamhannigan commented on src/features/game/events/landExpansion/feedBumpkin.ts:
```diff
@@ -22,25 +29,42 @@ export function feedBumpkin({ state, action }: Options): GameState {
-    throw new Error("You have none of this food type");
+  // throws error if there are not enough in the inventory to feed the bumpkin
+  const inventoryFoodCount = inventory[action.food] ?? new Decimal(0);
+  if (inventoryFoodCount.lessThan(action.amount)) {
```
```suggestion
  if (inventoryFoodCount.lessThan(feedAmount)) {
```

adamhannigan reviewed (approved):
Backend added ✅ 

Awesome work @elias7905 🚀
